### PR TITLE
ENH: Improve subject hierarchy performance

### PR DESCRIPTION
**Pull request by lassoan**
GetItemByDataNode was slow because it had to search the whole tree for an item that has a specific MRML node associated with it.

Made it faster by adding a vtkMRMLNode* to vtkSubjectHierarchyItem* cache.

### CODE REVIEW

**Comment by lassoan** [comment]
@cpinter can you have a look?

File: Libs/MRML/Core/vtkMRMLSubjectHierarchyNode.cxx [hunk 1]
```diff
@@ -709,30 +715,36 @@ vtkSubjectHierarchyItem* vtkSubjectHierarchyItem::FindChildByID(vtkIdType itemID
     {
     return itemIt->second;
     }
-  else
-    {
-    vtkWarningMacro("FindChildByID: Item cache does not contain requested ID " << itemID);
-    }
+  // It is not an error if item is not found in cache. It happens normally when
+  // scene has just been closed and widgets are updating themselves and trying to look up their selected item.
```
Comment (cpinter) [inline Libs/MRML/Core/vtkMRMLSubjectHierarchyNode.cxx:719]: Great, hopefully this will get rid of all those annoying error messages.

**Review by cpinter — APPROVED** [review]
Sorry, I did have a look yesterday and it seemed good. Thanks a lot!

**Comment by lassoan** [comment]
Thanks for the review.

**Pull request merged**